riscv-smtdeleg-sstcfg.adoc

```diff
@@ -0,0 +1,98 @@
+|===
+
+[WARN]
+====
+_There has been much discussion of the best way to access counters in S-mode.  This method requires 1 new CSR (`stselect`), and creates a scenario where there are 2 levels of indirection (siselect routes sireg* to t*, and tselect routes t* to the state associated with the selected trigger)._
+
+_Also proposed was to simply allocate a range of `siselect` values that serve to select the trigger, bypassing `tselect` when accessing trigger state through `sireg*`.  This avoids the 2nd level of indirection, but artificially limits the number of triggers accessible in S-mode.  `tselect` supports up to 2^32^ triggers, while `siselect` would likely allocate only enough values to cover practical implementations (256?)._
+
+_A third approach was considered, akin to what is spec'd here but accessing `tselect` via `sireg5` instead of through a new `stselect` CSR.  While saving a CSR, this approach would require `sireg5` to continue to access `tselect` while the other `sireg*` registers shift the state they acccess based on the `tselect` value.  This approach seems more awkward for implementation._
+
+_Feedback on these options, or others, is welcome._
+====
+
+Similarly, when `vsiselect` holds XXX, the `vsireg*` registers provide access to `tdata*` and `tinfo` in the same manner as described above for `sireg*`.
+
+=== Trigger State Access Modifications
+
+`sireg*` and `vsireg*` provide access to a subset of the fields in the native trigger CSRs, and in some cases fields are relocated.  The field modifications that apply when trigger registers are accessed via `sireg*` or `vsireg*` are documented below.
+
+.Delegated Trigger CSR Field Modifications
+[options="header", cols="20%,30%,50%"]
+|===
+| Sdtrig Register | `sireg*` Field Modifications | `vsireg*` Field Modifications
+| `tdata1` | M is read-only 0 
+
+VS and VU are read-only 0
+
+| M is read-only 0
+
+S accesses `tdata1`.VS
+
+U accesses `tdata1`.VU
+
+VS and VU are read-only 0
+| `tdata2` | none | none
+| `tdata3` | none | MHVALUE and MHSELECT are read-only 0
+| `tinfo` | none | none
+|===
+
+When `tdata1.dmode` = 1 for the trigger with index _i_, the trigger is reserved for Debug Mode use.  In that case, when `tselect` holds _i_ and the hart is not in Debug Mode, writes to `sireg` and `sireg[234]` are ignored while `siselect` holds XXX, and writes to `vsireg` and `vsireg[234]` are ignored while `vsiselect` holds XXX.
+
+== Supervisor and Virtual Supervisor Trigger State Access Control
+
+While the privilege mode is M or S and `siselect` holds XXX, illegal instruction exceptions are raised for attempts to access `sireg5` or `sireg6`.  Similarly, while the privilege mode is M or S and `vsiselect` holds XXX, illegal instruction exceptions are raised for attempts to access `vsireg5`, or `vsireg6`.
+
+While the privilege mode is VS and `vsiselect` holds XXX, virtual instruction exceptions are raised for attempts to access `sireg5` (really `vsireg5`) or `sireg6` (really `vsireg6`).
```

Comment:
`vsiselect` holds XXX -> `siselect`  holds XXX?

Reply:
No, in VS-mode, accesses to sireg* are redirected to vsireg*, and accesses to siselect are redirected to vsiselect.  So a guest OS will attempt to access siselect/sireg*, but it will actually access vsiselect/vsireg*.

Reply:
I got your point. By "`vsiselect` holds XXX" you mean the real CSR holds XXX instead of the programming shadow CSR hold XXX. It is a little bit misleading here. Probably we could align the description with `sireg*`, like `siselect` (really `vsiselect`) holds XXX.

Reply:
The phrasing of "siregX (really vsiregX)" refers to software accesses, since VS-mode software will attempt to access siregX, but the hardware will direct the access to vsiregX.  When speaking of the state held by a register, it is vsiselect that must hold XXX, not siselect.  siselect may (likely does) hold some different value, written by the hypervisor or other more privileged software.  It's true that the value held by vsiselect may be the result of a VS-mode attempt to write siselect, but it could also be the result of a direct write to vsiselect by (H)S-mode or M-mode.  Since this behavior doesn't depend on the method of access, but rather the result, only the end value is referenced here.  Hope that makes sense.